sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp

```diff
@@ -595,10 +783,24 @@ Azure::Core::_internal::UniqueHandle<HINTERNET> WinHttpTransport::CreateConnecti
   return rv;
 }
 
-Azure::Core::_internal::UniqueHandle<HINTERNET> WinHttpTransport::CreateRequestHandle(
+void _detail::WinHttpRequest::EnableWebSocketsSupport()
+{
+#pragma warning(push)
+  // warning C6387: _Param_(3) could be '0'.
+#pragma warning(disable : 6387)
+  if (!WinHttpSetOption(m_requestHandle.get(), WINHTTP_OPTION_UPGRADE_TO_WEB_SOCKET, nullptr, 0))
+#pragma warning(pop)
+  {
+    GetErrorAndThrow("Error while Enabling WebSocket upgrade.");
```

Comment:
`GetErrorAndThrow` has a defaulted `errorCode` parameter which is defaulted to `GetLastError()`. Everywhere you see `GetErrorAndThrow`, you can assume that the error code is included in the output.

Originally `GetErrorAndThrow` used to call `GetLastError()` internally (which is stupidly fragile), for the initial websocket work, I changed it to make the error code parameter a new parameter defaulted to `GetLastError` (because I didn't want to change the existing signature).

Reply:
> Originally `GetErrorAndThrow` used to call `GetLastError()` internally (which is stupidly fragile)...

For my edification, can you please explain why it is fragile?

Reply:
When GetErrorAndThrow calls GetLastError to discover the error code, it means that it can only work if the caller has called SetLastError before calling GetErrorAndThrow.

That is extremely fragile because Win32 errors can come from functions which don't call `SetLastError`. By making the error code a parameter (defaulted to `GetLastError`), it means that GetErrorAndThrow can be used both when a Win32 API returns a status code via SetLastError AND when the Win32 API explicitly returns the status code.

See also [this](https://www.youtube.com/watch?v=1NqmvFq016Y) for more context on the `SetLastError` vs return and error code issue.